hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java

```diff
@@ -341,7 +341,8 @@ private RedefinedWritableStates applyStateDefinitions(
                                             new VirtualMap<>(label, keySerializer, valueSerializer, dsBuilder);
                                     return virtualMap;
                                 },
-                                virtualMap -> virtualMap.registerMetrics(metrics));
+                                virtualMap -> virtualMap.registerMetrics(
+                                        stateRoot.getMetrics() != null ? stateRoot.getMetrics() : metrics));
```

Comment:
Why this change?

Reply:
`PlatformStateStore` and `EntityIdStateStore` are two peculiar (let's call them pre-loaded state stores) states stores which must be initialized before other stores in the `StoreFactory`s are initialized (although those two are again re-initialized in the `StoreFactory`s). As at the time those two are initialized, this instance of `Metrics` is still `UnavailableMetric`s.

The effect of this is that there's been an inherent limitation where if either of those pre-loaded state stores had a virtualmap, this line of code would have failed. We've gotten away till date because neither `PlatformStateStore` not `EntityIdstateStore` had a virtual map so this code was never called during that early initialization.

Now, I'm introducing another pre-loaded state store called `RosterStateStore`, which has a virtualmap. The right thing to do is pass the instance of `Metrics` that's been created into here. The least invasive way to do that is to set it in the state (the state already had a metrics field) and use that instead.

Reply:
Adding more context here. There's a chicken and egg problem at play where `PlatformStateService` and `EntityIdService` are needed to initialize state and do migrations, but they both also themselves need to be migrated. So what the existing code does is run migrations twice - the first for those two services only, and the second round for all services (including those two).

During the migration, `VirtualMap` registers an instance of metrics. Metrics is initially defaulted to `UnavailableMetrics` instance, and eventually replaced with the DefaultPlatformMetrics created earlier. The code path that sets the metrics have not been travelled before in the early round of migrations (because neither PlatformState nor EntityIdState had `VirtualMap`s) so this problem was always lying there. 

With the introduction of Roster schema as part of the first set of migrations, the metrics still carries the `UnavailableMetrics` instance and tries to register it which is unsupported (throws an exception). To avoid this problem, a concrete Metrics instance is required. The state (`MerkleStateRoot`) already carries the metrics, so I just bubbled the metrics already created by `ServicesMain` into the `MerkleStateRoot` early, so it's available before the first round of migrations. This line is saying to use the stateRoot metrics if set, or default to what was there before (for non-MerkleStateRoot states).

Now, as to whether the metrics is still needed or not, I spoke to @artemananiev before making this change and from what I gathered, we do need to make sure metrics are registered both when a new virtualMap is created or an existing one found. This is one of those.

Perhaps there's a second option of undeprecating `NoOpMetrics` and using it instead of `UnavailableMetrics`?

Reply:
Thank you for the explanation, it makes more sense now. Adding some of this context to the code in the form of comments would be helpful.